crypto/init.c

```diff
@@ -491,43 +499,61 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
      */
 
     if (stopped) {
-        if (!(opts & OPENSSL_INIT_BASE_ONLY))
+        if (opts != 0)
             CRYPTOerr(CRYPTO_F_OPENSSL_INIT_CRYPTO, ERR_R_INIT_FAIL);
         return 0;
     }
 
-    /*
-     * When the caller specifies OPENSSL_INIT_BASE_ONLY, that should be the
-     * *only* option specified.  With that option we return immediately after
-     * doing the requested limited initialization.  Note that
-     * err_shelve_state() called by us via ossl_init_load_crypto_nodelete()
-     * re-enters OPENSSL_init_crypto() with OPENSSL_INIT_BASE_ONLY, but with
-     * base already initialized this is a harmless NOOP.
-     *
-     * If we remain the only caller of err_shelve_state() the recursion should
-     * perhaps be removed, but if in doubt, it can be left in place.
-     */
-    if (!RUN_ONCE(&base, ossl_init_base))
+    if (!RUN_ONCE(&init_lock_once, do_init_lock)) {
+        CRYPTOerr(CRYPTO_F_OPENSSL_INIT_CRYPTO, ERR_R_INIT_FAIL);
         return 0;
+    }
 
-    if (opts & OPENSSL_INIT_BASE_ONLY)
-        return 1;
+    /* Base init by default, always */
+    opts |= OPENSSL_INIT_BASE;
 
     /*
-     * Now we don't always set up exit handlers, the INIT_BASE_ONLY calls
-     * should not have the side-effect of setting up exit handlers, and
-     * therefore, this code block is below the INIT_BASE_ONLY-conditioned early
-     * return above.
+     * We want two be able to pass the settings to RUN_ONCE functions; this
+     * requires a lock, since RUN_ONCE doesn't pass an extra argument.
+     * We want to be able to have certain things happen by default (loading
+     * the config file, for example).
+     * We want to be able to call OPENSSL_init_crypto() recursively (or at
+     * least let it happen because some functions called down the line from
+     * here may as well be called directly from the application, so we want
+     * to ensure initialization in both cases).
+     *
+     * Unfortunately, combining all three may lead to nasty lock ups,
+     * because a recursive call may try to lock the same lock that was
+     * already locked.
+     *
+     * Our solution for this problem is to keep track of what things are
+     * already being initialized and should therefore not be initialized
+     * again.  We do this by saving away what options have already been
+     * chosen in an earlier call, and making opts only hold the bits that
+     * weren't already there.
+     *
+     * Note that this really makes any init option a one chance only, even
+     * if one of the initializations fail.
      */
+    {
+        CRYPTO_THREAD_write_lock(init_lock);
+        opts &= ~last_opts;      /* Mask away what's already been done */
+        last_opts |= opts;       /* Add what we're currently going to do */
+        CRYPTO_THREAD_unlock(init_lock);
```

Comment:
This is going to do the wrong thing in multi-threaded code. Thread 1 gets to this code first and removes stuff from last_opts that its about to do and then gets de-scheduled before it actually initialises them. Thread 2 then gets initialised gets to here and sees that it doesn't need to initialise things and returns. It then continues expecting those things to have been initialised and fails.